src/msw/joystick.cpp

```diff
@@ -37,8 +37,127 @@
 
 #include <regstr.h>
 
+// Use optimised count trailing zeros where available.
+#ifdef __GNUC__
+    #define ctz(x) __builtin_ctz(x)
+#else
+int ctz(unsigned x) {
+   int n;
+
+   if (x == 0) return(32);
```

Comment:
This is wrong, `__builtin_ctz(0)` behaviour is undefined, so it can't be called with 0. Either the caller needs to check for this, or `ctz()` above must do it too.

Also, hardcoding 32 is not ideal, even if fine in practice.

Reply:
OK. How about throwing an exception in either case if the input is zero? What would the preferred exception be?

```C++
#ifdef __GNUC__
int ctz(usigned x)
{
    // throw some exception if x == 0
   return __builtin_ctz(x);
}
#else
int ctz(unsigned x)
{
  // throw some exception of x == 0
   int n;
   n = 1;
   if ((x & 0x0000FFFF) == 0) {n = n +16; x = x >>16;}
   if ((x & 0x000000FF) == 0) {n = n + 8; x = x >> 8;}
   if ((x & 0x0000000F) == 0) {n = n + 4; x = x >> 4;}
   if ((x & 0x00000003) == 0) {n = n + 2; x = x >> 2;}
   return n - (x & 1);
}
#endif
```

Reply:
How about this?

```C++

int ctz(unsigned x)
{
    wxCHECK_MSG(x > 0, 0, "Undefined for x == 0.");
#ifdef __GNUC__
   return __builtin_ctz(x);
#else
   int n;
   n = 1;
   if ((x & 0x0000FFFF) == 0) {n = n +16; x = x >>16;}
   if ((x & 0x000000FF) == 0) {n = n + 8; x = x >> 8;}
   if ((x & 0x0000000F) == 0) {n = n + 4; x = x >> 4;}
   if ((x & 0x00000003) == 0) {n = n + 2; x = x >> 2;}
   return n - (x & 1);
#endif
}
'''

Reply:
Looks good, thanks.